### PR TITLE
fix(action): stop promoting NEW-only compile/source/debug blocks pair-wide

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

Follow-up to #1159 (merged at `57f426dd5`, before this fix was pushed — see [this reply](https://github.com/abicheck/abicheck/pull/1159#discussion_r3965063772) on the original PR). Carries forward correctness fixes for the sources-root config-discovery merge in `action/run.sh`'s `add_compile_context_flags()`, refined across two review rounds on this PR itself.

## Motivation

`add_compile_context_flags()` discovers a `--sources` tree's own `.abicheck.yml` and replaces some of the checkout-root document's blocks with it before writing the merged result as the single `--config` this Action passes to `abicheck`. Two Codex review rounds found the promoted-block list was wrong in opposite directions:

1. **First round:** `compile:`/`source:`(singular)/`debug:` are **pair-wide** for single-pair `compare` — `resolve_compile_context()` applies `compile:` "to both sides", and `resolved_cfg.source_method`/`debug_format` are each resolved once for the whole comparison. `embed_build_source()` (the function this logic was modeled on) never reads any of those three at its `--sources`-anchored call site — only `build.query`/`compile_db`/`targets` and `sources.public_headers`/`exclude`/`graph`. Promoting them from a NEW-side-only `--sources` tree's config into the single shared `--config` silently applied NEW-only settings to OLD's own header parsing too.
2. **Second round:** excluding those three blocks *unconditionally* (the first round's fix) was correct for pairwise `compare` but wrong for `dump`/`scan --against`, which have no "other side" for them to leak into (`dump` has one operand; `scan`'s own `-H`/`-I` have always applied only to the scanned ARTIFACT). `merge_compile_config()`'s own docstring confirms the `--sources` tree's config IS the intended, only source of `compile:` for that one operand there — mirroring `embed_build_source()`'s own single-sided selection.

## Changes

- Narrowed the sources-root block-replacement list to `("build", "sources")` always, plus `compile`/`source`/`debug` **conditionally** — included only when the caller's `$MODE` is NOT `compare` (single-pair).
- Added `_compile_context_sources_pairwise()` (keyed on `$MODE`) and a 6th parameter on `_merge_config_overlay_with_discovered_project_config` to thread this decision through.
- `found_path` is reassigned to the sources root exactly when `compile:` was actually sourced from it (single-sided callers), so `compile.include_dirs` keeps anchoring correctly either way.
- Test coverage for both modes: a dump-mode test proving `compile:`/`source:`/`debug:` ARE promoted there, a compare-mode test proving they are NOT, and matching empty-sources-root companions for each mode.
- `architecture/debt.yaml`: baseline bump for `tests/test_action_compile_context_parity.py` (1593 → 1745, across both rounds) with rationale.

## Bug-fix test contract

- Bug class: a shared config-merge helper applying a setting's correct per-side/per-mode scope inconsistently — first promoting a per-side-only setting pair-wide, then (correcting that) excluding a genuinely single-sided setting from the one mode that needs it.
- Publicly observable failure: (a) a GitHub Action `compare` run with a compile-context input plus a `sources` tree carrying its own `compile:`/`source:`/`debug:` block would silently apply that NEW-only setting to OLD's header parsing too; (b) after the first fix, the same input combination under `dump`/`scan --against` would silently *lose* the sources tree's own `compile:`/`debug:` settings instead of using them, changing what compile context is used to parse dump's own headers.
- Regression test fails on base: `test_sources_root_compile_block_is_never_sourced_from_it_under_pairwise_compare` (fails against the pre-first-fix code) and `test_sources_root_compile_block_is_sourced_from_it_under_single_sided_dump` (fails against the code between the two fixes, where dump had lost this).
- Negative control: `test_sources_root_build_block_is_discovered_and_merged`/`test_sources_root_sources_block_is_also_carried` (dump mode) continue to pass unchanged — the two genuinely mode-independent blocks (`build:`/`sources:`) are unaffected by either fix.
- Public-surface test: yes — invoked through the real `add_compile_context_flags()` shell function via `_run_region_with_cwd`, which extracts and executes the actual script region under a real `$MODE` value, asserting on the real JSON `--config` overlay it writes.
- Axes covered: both modes (`compare` pairwise, `dump` single-sided) × both scenarios (a populated sources-root config, an empty one) — 4 tests total for the `compile:` axis alone.
- General invariant: a NEW-side-only sources-root config's blocks may only be promoted into the single shared `--config` when the resolved value would actually apply to that same, single operand under the caller's mode — never when the mode also resolves that block pair-wide across two independently-parsed operands.
- Malicious fixture + side-effect absence: every test in `TestCompileContextDiscoversSourcesRootOwnConfig` writes a real, untrusted (repo-discovered, not operator-authored) `--sources`-tree `.abicheck.yml` fixture to disk and runs it through `_run_region_with_cwd`, which executes the actual extracted `add_compile_context_flags`/`_merge_config_overlay_with_discovered_project_config` shell functions and their real Python subprocess (not a mock, and not a text-only assertion of the fixture's own YAML). `test_sources_root_compile_block_is_never_sourced_from_it_under_pairwise_compare` plants a hostile-shaped sources-root config (`compile.sysroot: /opt/new-side-only-sysroot`, `source.method: s6`, `debug.format: dwarf`) under pairwise `compare` mode and asserts, against the *real, executed* merged `--config` JSON, that none of those three keys crossed into the pair-wide document (`doc["compile"]["sysroot"]` stays the checkout-root's own value; `source`/`debug` are absent entirely) — the side effect this fix exists to prevent (a NEW-only setting silently reaching OLD's header parsing) is asserted absent from the real artifact the script produced, not merely from the input fixture's own text. `test_empty_sources_root_config_clears_the_checkout_root_blocks` and its pairwise companion do the same for the empty/non-mapping-fixture edge case (`yaml.safe_load` returning `None`), which is the other place a hostile-but-degenerate input could otherwise slip past a text-only check.

## Checklist

- [x] Tests added / updated for new behaviour
- [ ] Changelog fragment added (`scriv create`) if this touches `abicheck/**/*.py` — see `changelog.d/README.md` (N/A — only `action/run.sh`, `tests/`, `architecture/debt.yaml` touched)
- [ ] Docs updated if user-visible behaviour changed (N/A — internal Action config-merge fix, no public interface change)
- [x] `pre-commit run --all-files`-equivalent checks (`bash -n`, `check_architecture.py`) pass locally
- [x] No new `mypy` or `ruff` errors introduced

## Verification

- `bash -n action/run.sh` — clean.
- `pytest tests/test_action_compile_context_parity.py` — 49 passed.
- Broader sweep `pytest tests/ -k "action_run_sh or action_compile_context or action_release_topology"` — 476 passed.
- `ruff check` — clean.
- `python scripts/check_architecture.py` — 0 errors.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01GD2nKTWjYyh4zKeWLgPtCP)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected configuration handling during compile-context discovery so source-root settings are applied appropriately across comparison, dump, and scan workflows.
  * Preserved pairwise comparison context when comparing against a live library, while continuing to apply full configuration when using stored snapshots.
  * Improved recognition of stored snapshots based on file content, including binaries with misleading filenames and compressed or JSON snapshots.

* **Tests**
  * Expanded coverage for mode-specific configuration behavior and snapshot detection.

### CODE REVIEW

File: action/run.sh [hunk 1]
```diff
@@ -702,29 +716,30 @@ else:
...
-            # both are read by embed_build_source()'s BuildConfig and
-            # both belong to whichever ONE file backs it (Codex review,
-            # fresh evidence: a source-root sources.graph: full silently
-            # narrowing to the checkout-root's/default summary before
-            # collect_inline_pack() reads cfg.graph_detail).
-            for _blk_key in ("build", "sources", "compile", "source", "debug"):
+            # DISTINCT top-level block from "source" (singular, pair-wide
+            # source.method -- deliberately excluded, see above). Only the
+            # two genuinely NEW-side-scoped blocks are replaced here.
+            for _blk_key in ("build", "sources"):
```
[the remark this answers is not in the document]
Comment (napetrov) [reply action/run.sh]: ---
Fixed in dbb059fd9.

Confirmed exactly as described: `merge_compile_config()`'s own docstring already states the `--sources` tree's config is the intended, only source of `compile:` for a single operand (mirroring `embed_build_source()`'s own single-sided `build:`/`sources:` selection) — my prior fix excluded `compile:`/`source:`(singular)/`debug:` unconditionally, which was correct for pairwise single-pair `compare` but wrong for `dump`/`scan --against`, both of which have no "other side" for those settings to leak into (`dump` has one operand; `scan`'s own `-H`/`-I` have always applied only to the scanned ARTIFACT, never to `--against`).

Made the exclusion mode-aware: a new `_compile_context_sources_pairwise()` helper (keyed on `$MODE`) and a 6th parameter on `_merge_config_overlay_with_discovered_project_config` gate whether `compile:`/`source:`/`debug:` are included in the sources-root block-replacement list — excluded only when `$MODE == "compare"` (single-pair), included otherwise (`dump`, `scan`). `found_path` is now reassigned to the sources root exactly when `compile:` was actually sourced from it, so `compile.include_dirs` keeps anchoring correctly either way.

Test changes: split the empty-sources-root test into a dump-mode version (`compile:` cleared, matching single-sided promotion) and a new pairwise-compare-mode companion (`compile:` survives); moved the compile-block-never-sourced test from dump mode (where it was now testing the wrong mode) to compare mode; added a new dump-mode test proving `compile:`/`source:`/`debug:` ARE promoted there. All adversarially confirmed against the pre-fix code.

`tests/test_action_compile_context_parity.py` (49 tests), the broader `action_run_sh`/`action_compile_context`/`action_release_topology` sweep (476 tests), `ruff check`, and `check_architecture.py` (0 errors, `debt.yaml` baseline bumped 1653 → 1745) all pass/clean.
---

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
**CI failure not caused by this PR's own diff, fixed anyway (fixed in 385aa2c70).**

`unit-tests (macos-latest, 3.13, false)` failed on `dbb059fd9` with:
```
tmphftjn17y.sh: line 716: CMD: bad array subscript
tmphftjn17y.sh: line 716: CMD[-1]: unbound variable
```

Root cause: `tests/test_action_release_topology_config.py::TestReleaseTopologyOverlayCleansUpOnExit::test_release_topology_overlay_is_removed_after_the_trap_fires`'s bash harness used `${CMD[-1]}` to read the last element of the `CMD` array — negative array subscripts are a bash 4.3+ feature. macOS's stock bash is 3.2 (per `action/AGENTS.md`'s own portability requirement), which raises `bad array subscript` on that syntax. Confirmed identical on `origin/main` (the same line, unmodified) — this is not something either of this PR's own commits introduced.

Since it was blocking this PR's own macOS unit-test lane and the fix is a trivial, safe one-line portability correction (`${CMD[${#CMD[@]}-1]}`, portable since bash 3.0), I fixed it directly rather than leaving the lane red. `pytest tests/test_action_release_topology_config.py` (34 tests) passes; `ruff check`/`check_architecture.py` clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by napetrov** [comment]
**Another CI failure not caused by this PR's own diff, fixed anyway (fixed in 028fab7c5).**

`unit-tests (windows-latest, 3.13, false)` failed on `385aa2c70` with:
```
FAILED tests/test_action_release_topology_config.py::TestReleaseTopologyOverlayPreservesWindowsQualifiedBuildConfigPath::test_same_drive_letter_shaped_path_is_prefixed_on_non_windows
FAILED tests/test_action_release_topology_config.py::...::test_ordinary_relative_path_still_gets_pwd_prefix_on_windows
```

Root cause: both tests built their expected `$PWD/`-prefixed value from Python's `str(tmp_path)`, silently assuming that always matches bash's own `$PWD` builtin for the same directory. That coincidentally holds on Linux/macOS (where these tests were originally verified — no Windows toolchain was available at the time) but not on a real Windows runner: there, `bash` is Git Bash/MSYS, whose `$PWD` is always its own POSIX-style rendering (`/c/Users/...`) regardless of the test's forced `$OSTYPE=linux-gnu` — that variable only steers which branch the *code under test* takes, not bash's own real `$PWD`. Confirmed identical on `origin/main` (both tests, unmodified) — not something either of this PR's commits introduced.

Fixed by having the test harness also capture bash's own observed `$PWD` for the given `cwd` and building both expected values from that real, observed value instead of `str(tmp_path)` — portable by construction rather than by coincidence of which host runs it. `pytest tests/test_action_release_topology_config.py` (34 tests) passes; broader `action_run_sh`/`action_compile_context`/`action_release_topology` sweep (476 tests), `ruff check`, `check_architecture.py` (0 errors, new `debt.yaml` entry for this file crossing the 1200-line test cap) all clean.

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: action/run.sh [hunk 2]
```diff
@@ -903,6 +958,25 @@ _rm_overlay_on_early_exit() {
...
+# resolved compile context. `dump` has one operand; `scan --against`'s own
+# -H/-I have always applied only to the scanned ARTIFACT, never to
+# --against (never live headers on that side at all) -- neither has an
+# "other side" a --sources tree's own compile:/debug: could leak into, so
+# both stay single-sided. The release-style directory/package `compare`
+# fan-out never calls add_compile_context_flags at all (it rejects every
+# compile-context input outright), so it never reaches this helper.
+_compile_context_sources_pairwise() {
+  if [[ "$MODE" == "compare" ]]; then
+    echo "pairwise"
```
[the remark this answers is not in the document]
Comment (napetrov) [reply action/run.sh]: Fixed in 10acef771.

Confirmed exactly as described: `_compile_context_sources_pairwise()`'s `$MODE == "compare"` check alone couldn't distinguish a genuinely live-pair compare from a compare where OLD is a stored snapshot/resolved `--abi-baseline` and only NEW is live with `--sources` — the latter has no "other side" for those settings to leak into, same as `dump`/`scan --against`.

Added `_old_library_is_stored_snapshot()`, an extension-based check on `$INPUT_OLD_LIBRARY` (`*.json`/`.json.gz`/`.json.zst`) mirroring this script's own existing `[[ "${INPUT_AGAINST:-}" == *.json ]]` snapshot check and the `*.abicheck.json`/`.gz`/`.zst` glob triple already used twice for baseline-asset discovery — `resolve_input()` itself only ever content-sniffs (magic bytes), but that's not cheaply reproducible in bash ahead of the CLI, and this extension check is the bar this script already relies on for the identical class of decision elsewhere. By the time `add_compile_context_flags` runs, `$INPUT_OLD_LIBRARY` already reflects an `--abi-baseline` auto-fetch (it's assigned `$BASELINE_FILE`, itself always a downloaded `*.abicheck.json[.gz|.zst]`), so this one check covers both the auto-fetched-baseline and directly-named-snapshot cases.

`_compile_context_sources_pairwise()` is now pairwise only when `$MODE == "compare"` **and** OLD is not a stored snapshot.

Deliberately narrow scope: an ABICC Perl-dump old-library is also stored/live-extraction-free but has no comparably reliable extension convention, so it isn't detected here and stays on the pre-existing, safe "pairwise" side — not a generalized live/stored classifier, just this one already-precedented signal.

Five new tests cover the fix (`test_sources_root_compile_block_is_sourced_from_it_under_compare_with_stored_old_snapshot`, parametrized over `.abicheck.json`/`.json`/`.JSON`/`.json.gz`/`.json.zst`) plus a negative control (`..._stays_pairwise_when_old_library_is_a_live_binary_shaped_like_json`) proving a live-binary old-library still stays pairwise.

`tests/test_action_compile_context_parity.py` (55 tests) + `tests/test_action_release_topology_config.py` (89 total), the broader `action_run_sh`/`action_compile_context`/`action_release_topology` sweep (482 tests), `ruff check`, and `check_architecture.py` (0 errors, `debt.yaml` baseline bumped 1806 → 1903) all pass/clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: action/run.sh [hunk 3]
```diff
@@ -903,6 +958,63 @@ _rm_overlay_on_early_exit() {
...
+# (also stored, also live-extraction-free) has no comparably reliable
+# extension convention to key off, so it is not detected here and stays on
+# the prior, safe-by-construction "pairwise" side -- not a generalized
+# live/stored classifier, just this one already-precedented signal.
+_old_library_is_stored_snapshot() {
+  local lower
+  lower=$(printf '%s' "$1" | tr '[:upper:]' '[:lower:]')
+  case "$lower" in
+    *.json | *.json.gz | *.json.zst)
+      return 0
```
[the remark this answers is not in the document]
Comment (napetrov) [reply action/run.sh]: Fixed in 8eeb80281.

Confirmed exactly as described — good catch, and correct that the negative test only exercised `old.so`, never the resolver-order case that actually mattered.

Replaced the extension-only `_old_library_is_stored_snapshot()` with content sniffing in the same order `resolve_input()` itself uses: native-binary magic bytes are checked **first** (ELF `7f454c46`, PE `4d5a`, and all four Mach-O/fat-binary magic variants in both byte orders, via the same `od -An -tx1` idiom `_is_release_style_operand()` already uses), and only if none of those match does it fall through to gzip/zstd magic or a bare `{` prefix for a JSON snapshot. A path that can't be opened at all (doesn't exist yet) falls back to the pre-existing safe "pairwise" default, mirroring `_is_release_style_operand`'s own `-f` fallback.

Rewrote the prior extension-only parametrized tests to write real bytes carrying each case's own magic prefix (rather than relying on a suggestively-named, empty path), and added three tests:
- the exact case you named: a live ELF binary literally named `old.json` — asserts it stays pairwise.
- a plain live-binary negative control with real ELF magic bytes (replacing the old `old.so`-with-no-content version).
- a does-not-exist-on-disk fallback case — asserts pairwise.

`tests/test_action_compile_context_parity.py` (59 tests) + `tests/test_action_release_topology_config.py` (93 total), the broader `action_run_sh`/`action_compile_context`/`action_release_topology` sweep (484 tests), `ruff check`, and `check_architecture.py` (0 errors, `debt.yaml` baseline bumped 1903 → 1991) all pass/clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (napetrov) [reply action/run.sh]: Correction to the test counts in my previous reply: `tests/test_action_compile_context_parity.py` is 57 tests (not 59), and combined with `tests/test_action_release_topology_config.py` it's 91 total (not 93) — both still pass. Everything else in that reply stands.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: action/run.sh [hunk 4]
```diff
@@ -903,6 +958,88 @@ _rm_overlay_on_early_exit() {
...
+  esac
+  case "$magic4" in
+    1f8b*) return 0 ;;      # gzip (a compressed *.json.gz snapshot)
+    28b52ffd) return 0 ;;   # zstd (a compressed *.json.zst snapshot)
+  esac
+  # Bare JSON snapshot: first byte is '{' (mirrors sniff_text_format()'s
+  # own `head.startswith("{")` for the uncompressed case).
+  local first_byte
+  first_byte=$(od -An -c -N 1 "$path" 2>/dev/null | tr -d ' \n')
+  [[ "$first_byte" == "{" ]]
```
[the remark this answers is not in the document]
Comment (napetrov) [reply action/run.sh]: Fixed in 48cf3b81c.

Confirmed exactly as described. Rather than adding a third format-specific case (leading-whitespace JSON, then an ABICC Perl dump, and whatever the next round would have found), I collapsed `_old_library_is_stored_snapshot()` to its actual invariant: NOT a recognized live-binary format (ELF/PE/Mach-O magic bytes) IS a stored operand, full stop. `resolve_input()` checks binary magic first, and every non-binary shape it goes on to accept (JSON with or without leading whitespace, compressed JSON, an ABICC Perl dump, a raw BTF/CTF blob, symvers, or simply unrecognized content) is equally extraction-free for this decision — none of them are parsed via `-H`/`-I`/`ast-frontend`/a `compile:` block at all, so there's no need to keep enumerating each format's own sniffing rule separately.

Three new parametrized cases (leading-whitespace JSON, an ABICC Perl dump, and unrecognized content) prove the collapsed rule handles all three without a dedicated branch, adversarially confirmed against the pre-fix (seventh-round) code.

Also split the pairwise-vs-single-sided-OLD test coverage (6 tests, now 13 with the new cases) into a new sibling module, `tests/test_action_compile_context_old_library_liveness.py` — the added cases pushed `tests/test_action_compile_context_parity.py` past the AI-readiness file-size gate's 2000-line hard cap, so this mirrors `tests/test_action_release_topology_windows_paths.py`'s own identical split.

`tests/test_action_compile_context_parity.py` (47 tests) + `tests/test_action_compile_context_old_library_liveness.py` (13 tests) + `tests/test_action_release_topology_config.py`/`tests/test_action_release_topology_windows_paths.py` (96 total across all four), the broader `action_run_sh`/`action_compile_context`/`action_release_topology` sweep (596 tests), `ruff check`, and `check_architecture.py` (0 errors) all pass/clean.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: action/run.sh [hunk 5]
```diff
@@ -947,6 +1002,88 @@ _rm_overlay_on_early_exit() {
...
+# fifth round: the earlier `$MODE == "compare"` check alone was too coarse
+# and silently discarded a live NEW-with-`--sources` operand's own
+# compile:/source:/debug: settings whenever OLD was a stored/baseline
+# snapshot) -- none of these three shapes has an "other side" a --sources
+# tree's own compile:/debug: could leak into, so all three stay
+# single-sided. The release-style directory/package `compare` fan-out never
+# calls add_compile_context_flags at all (it rejects every compile-context
+# input outright), so it never reaches this helper.
+_compile_context_sources_pairwise() {
+  if [[ "$MODE" == "compare" ]] && ! _old_library_is_stored_snapshot "${INPUT_OLD_LIBRARY:-}"; then
```
[the remark this answers is not in the document]
Comment (napetrov) [reply action/run.sh:1074]: Fixed in f9682bce7.

Confirmed exactly as described. `_compile_context_sources_pairwise()` was keyed purely on `$MODE` (pairwise only for `compare`, always single-sided for `scan`), but `scan --against` routed through `compare` internally (ADR-068 D2) is structurally the same shape: a live `--against` undergoes real header/debug extraction under the same shared compile context as the scanned artifact, exactly like `compare`'s own OLD operand.

Generalized the decision to resolve the OLD-like operand per mode (`$INPUT_OLD_LIBRARY` for `compare`, `$INPUT_AGAINST` for the translated `scan` branch) rather than assuming single-sided for every non-`compare` mode:
- `dump` keeps its own unconditional single-sided shape (it genuinely has no OLD-like operand at all).
- The *legacy* `scan` CLI branch (`$_SCAN_NEEDS_LEGACY_CLI == "true"`) never reaches this helper — it forwards these flags literally as its own CLI options and is unaffected.
- The translated `scan` branch now shares the exact same `_old_library_is_stored_snapshot()` classifier as `compare`, applied to `$INPUT_AGAINST`.

Two new tests in `tests/test_action_compile_context_old_library_liveness.py` cover the translated-scan branch directly (live `--against` → pairwise/excluded; stored-baseline `--against` → single-sided/promoted), extracting that branch's own region via a distinct header-comment anchor (its bare `elif [[ "$MODE" == "scan" ]]; then` text is now also referenced inside this function's own updated docstring, so anchoring on that substring directly would have matched the wrong, non-executable location).

`tests/test_action_compile_context_old_library_liveness.py` (15 tests) + `tests/test_action_compile_context_parity.py` (47 tests), the broader `action_run_sh`/`action_compile_context`/`action_release_topology`/`action_scan` sweep (598 tests), `ruff check`/`ruff format --check`, and `check_architecture.py` (0 errors) all pass/clean.

---
_Generated by [Claude Code](https://claude.ai/code)

---
_Generated by [Claude Code](https://claude.ai/code)_
